### ISSUE TITLE
Looser return types for before/afterEach and non-async functions with asyncProperty

### ISSUE THREAD

**Issue by sommd** [dubzzz/fast-check#480]
## 🚀 Feature Request

This is sort-of a two part feature request:

1. It would be nice if the return type of the `beforeEach` and `afterEach` hooks didn't need to return `void`, and instead could return `unknown`. This would be similar to what [Jest does](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L322).
2. It would be nice if any of the functions of `asyncProperty` (the predicate, beforeEach hook and afterEach hook) didn't have to be async.

## Motivation

Both of these features would reduce the amount of code that needs to be written and improve the usability of the library.

For part 1, the main motivation is so that simple before/after hooks can be one line:

```ts
// This:
property(
  // ...
).afterEach(() => jest.clearAllMocks());

// Instead of this:
property(
  // ...
).afterEach(() => { // Needs braces as the return type must be void
  jest.clearAllMocks();
});
```

The motivation for part 2 is similar, but is only really applicable if part 1 is implemented too:

```ts
// This:
asyncProperty(
  // ...
).afterEach(() => jest.clearAllMocks());

// Instead of this:
asyncProperty(
  // ...
).afterEach(async () => {
  jest.clearAllMocks();
});
```

It's a small improvement, but it is nicer :slightly_smiling_face:.

Part 2 could also allow for a non-async predicate, while having async before/after each functions:

```ts
asyncProperty(
  fc.nat(),
  (x) => // Non-async predicate
).afterEach(async () => {
  // Some async code
});
```

There might not be many use cases for this though, but it would be easy to support.

**Comment by dubzzz** [comment]
@sommd Thanks for the issue, the `beforeEach` and `afterEach` not asking for a specific return type would clearly be a great. But I still have some questions about it. 

It seems to be easy to implement in `asyncProperty` without any breaking change - see https://github.com/dubzzz/fast-check/blob/master/src/check/property/AsyncProperty.generic.ts

Concerning `property`, letting the user able to return anything might be error prone as she/he might return promises that will never be awaited nor considered as type errors 🤔 That would be my main concern around this feature: asymmetry between async and sync properties.

Concerning point 2, I'd be interested to know in which cases you had such need: sync property with async afterEach or beforeEach.

**Comment by sommd** [comment]
Hmm, that's a good point about accidentally returning a Promise. You might be able to prevent this using a conditional type, something like this:

```ts
function property<T>(x: () => T): T extends Promise<unknown> ? never : Property;
```

Instead of `never` it could be a string type like `"Property returns a Promise. You probably meant to use asyncProperty instead."`, which would be incompatible but also give a (kinda hacky) error message.

Would it also be possible for `property`

**Comment by dubzzz** [comment]
@sommd Ok let's try to support Part 1 ;)

Your trick with never should work (maybe need to extend it to `PromiseLike`).
Do you want to try to do a PR?

If so please add a test checking the failing cases in https://github.com/dubzzz/fast-check/blob/master/test/type/main.ts
Something like: `fc.assert(fc.property(fc.nat(), _ => {}).beforeEach(async() => {}))` is not expected to compile but `fc.assert(fc.property(fc.nat(), _ => {}).beforeEach(() => {}))` or `fc.assert(fc.asyncProperty(fc.nat(), _ => {}).beforeEach(async() => {}))` should be fine ;)

**Comment by sommd** [comment]
OK, made a PR to implement part 1 and partially implement part 2 (only non-async before/afterEach).

I discovered that a function that returns anything (including `Promise`) is assignable to `() => void`, so it's already possible for a `property` to have an `async` before/afterEach (even the actual property can be `async`). So I decided not to make any changes for that in this PR, maybe we should open a separate issue to fix that?

The PR only changes `AsyncProperty` so that before/afterEach can return any value (`Promise` or non-promise).
